WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,443 @@
+		/// reissuance node's output edges always fully account for the
+		/// dependent requests, including the zero credentials required for
+		/// them. When the smaller nodes appear on the input side, the non-zero
+		/// values are sufficient to fill the reissuance node's in edge set and
+		/// requires only one edge to fully drain the (remaining) balance, so
+		/// there will be an extra zero valued credential (requested normally,
+		/// incl. range proof).</para>
+		///
+		/// <para>New reissuance nodes fully absorb the value of the nodes they
+		/// substitute with no additional dependencies required, so each one
+		/// reduces the bipartite graph problem to a smaller one (by K-1 == 1),
+		/// since the replaced nodes no longer need to be considered.</para>
+		///
+		/// <para>When the list of nodes has been reduced to the remaining
+		/// non-zero out degree of the largest magnitude node edges that cancel
+		/// out positive and negative values are added. This will fully
+		/// discharge the largest magnitude node, except when it is positive and
+		/// all of the remaining negative nodes on the graph add up to less than
+		/// it.</para>
+		///
+		/// <para>There are two related/composing special cases mainly affecting
+		/// vsize credentials - when the positive valued nodes with remaining
+		/// out degree > 1 have a uniform and sufficient balance to cover all of
+		/// the negative balances, instead of taking the largest node all of the
+		/// equal valued nodes are reduced together resulting in a more balanced
+		/// structure overall. This is combined with another special case that
+		/// checks if pairing positive and negative nodes in a 1:1
+		/// correspondence is possible after aggregation into reissuance nodes.
+		/// This can apply to amount credentials as well when consolidating
+		/// multiple equal valued inputs, but is only expected to regularly
+		/// occur for vbyte credentials.</para>
+		///
+		/// <para>After all negative value nodes have been discharged, the
+		/// remaining in-edges of all nodes must be filled with zero
+		/// credentials. These are added according to the graph order, by
+		/// extending new edges from nodes whose in-degree is already maximized
+		/// but whose out degree is not. This again deals only with a bipartite
+		/// graph, because reissuance nodes consolidating output nodes leave no
+		/// zero edges unaccounted for in the nodes they replace, whereas on the
+		/// input side the structure fans in so necessarily it leaves no nodes
+		/// with non-maximized in-degrees, it can only increase the available
+		/// out degree for nodes which are not fully consumed.</para></remarks>
+		private DependencyGraph ResolveCredentials()
+			=> CredentialTypes.Aggregate(
+				CredentialTypes.Aggregate(this, (g, credentialType) => g.ResolveNegativeBalanceNodes(credentialType)),
+				(g, credentialType) => g.ResolveZeroCredentials(credentialType));
```

Comment:
How about:

```cs
private DependencyGraph ForEachCredentialType(Func<DependencyGraph, CredentialType, DepedencyGraph>> operation)
    => CredentialTypes.Aggregate(this, operation);
```

and then:

```cs
private DependencyGraph ResolveCredentials() => this
    .ForEachCredentialType((g, credentialType) => g.ResolveNegativeBalanceNodes(credentialType))
    .ForEachCredentialType((g, credentialType) => g.ResolveZeroCredentials(credentialType));
```

or maybe even something like this (not sure if you can do this with instance methods, but if not maybe it makes sense to make some of this stuff static anyway, since they are functions on a type):

```
private DependencyGraph ResolveCredentials() => this
    .ForEachCredentialType(ResolveNegativeBalanceNodes)
    .ForEachCredentialType(ResolveZeroCredentials);
```

Reply:
It not the same that `this.Aggregate().Aggregate()`?

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/5878